### ISSUE TITLE
generating prediction-stat:"P" ,"R" , "mAP"  per image in test.py

### ISSUE THREAD

**Issue by tjbe2021** [ultralytics/yolov5#2437]
## Additional context
Hi there, I need some help to generate prediction-stats per image in the "Val" folder when run the test.py script, I couldn't find a way to do it by going through the code. Appreciate the help.

**Comment by maheshmechengg** [comment]
When i give image size other than model (416) to test.py my anchors misplace on test labels & test predict:
![image](https://user-images.githubusercontent.com/49720282/110890400-f9986c80-8315-11eb-8225-5dcc53b5e5fe.png)

**Comment by glenn-jocher** [comment]
@tjbe2021 test.py runs by default on your data.yaml `test: directory`. You can point it to the data.yaml `val: directory` by:
`python test.py --task val`

@maheshmechengg please raise a new issue for new topics. Your labels are incorrect.

**Comment by tjbe2021** [comment]
@glenn-jocher  Thanks for the prompt reply, I've changed the default to val, tI get the values for each class, and for all. 

However , I'm trying to generate  the P, R, mAP for the each image inside the val directory. So that  I could perform some sandy checks, is that possible with this ? I tried tweaking the code, but doesn't seems it's working.

**Comment by glenn-jocher** [comment]
@tjbe2021 AP is not computed per image, it is computed per class over all images, and then averaged as mAP. If you want to obtain mAP on one image, then your dataset would have to be 1 image.

**Comment by tjbe2021** [comment]
@glenn-jocher great thanks for the answer, that make sense! By any chance under metrics.py are you producing any results on FP,TP,  values?

**Comment by glenn-jocher** [comment]
@tjbe2021 YOLOv5 TP and FP vectors are computed here:
https://github.com/ultralytics/yolov5/blob/c8c5ef36c9a19c7843993ee8d51aebb685467eca/utils/general.py#L250-L319

They don't print out by default, you'd have to introduce some custom code to see them.

`fpc` and `tpc` are the FP and TP arrays of shape (n,10), for the 10 iou thresholds of 0.5:0.95. If you look at the last row, this is the FP and TP count per iou threshold:

```python
tpc.shape
Out[3]: (3444, 10)
fpc.shape
Out[4]: (3444, 10)
tpc[-1]
Out[5]: array([138, 124, 105,  91,  80,  66,  54,  38,  22,   9])
fpc[-1]
Out[6]: array([3306, 3320, 3339, 3353, 3364, 3378, 3390, 3406, 3422, 3435])
```

So at 0.5 iou and 0.001 confidence threshold, for class 0, dataset inference results in 138 TPs and 3306 FPs.

**Comment by glenn-jocher** [comment]
@tjbe2021 good news 😃! Your original issue may now be fixed ✅ in PR #5727. This PR explicitly computes TP and FP from the existing Labels, P, and R metrics:

```bash
TP = Recall * Labels
FP = TP / Precision - TP
```

These TP and FP per-class vectors are left in val.py for users to access if they want:
https://github.com/ultralytics/yolov5/blob/36d12a500eae4561d09d4955e1b50b12e57bf6c6/val.py#L240

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
